server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -0,0 +1,526 @@
+            TestClusterNode master = nodes.get(state.nodes().getMasterNode().getName());
+            assertNotNull(master);
+            assertTrue(master.node.isMasterNode());
+            return master;
+        }
+    }
+
+    private final class TestClusterNode {
+
+        private final DeterministicTaskQueue deterministicTaskQueue;
+
+        private final TransportService transportService;
+
+        private final ClusterService clusterService;
+
+        private final RepositoriesService repositoriesService = mock(RepositoriesService.class);
+
+        private final SnapshotsService snapshotsService;
+
+        private final SnapshotShardsService snapshotShardsService;
+
+        private final IndicesService indicesService;
+
+        private final IndicesClusterStateService indicesClusterStateService;
+
+        private final DiscoveryNode node;
+
+        private final MasterService masterService;
+
+        private final AllocationService allocationService;
+
+        private final AtomicReference<ClusterState> currentState = new AtomicReference<>();
+
+        private final NodeClient client;
+
+        private final NodeEnvironment nodeEnv;
+
+        /**
+         * Short circuit mock transport that simply invokes handlers on the discovery nodes directly.
+         */
+        private final MockTransport mockTransport = new MockTransport() {
+            @Override
+            protected void onSendRequest(long requestId, String action, TransportRequest request, DiscoveryNode node) {
+                assertFalse(TestClusterNode.this.node.equals(node));
+                final RequestHandlerRegistry<TransportRequest> handlerRegistry =
+                    testClusterNodes.nodes.get(node.getName()).mockTransport.getRequestHandler(action);
+                try {
+                    SetOnce<TransportResponse> responseSetOnce = new SetOnce<>();
+                    TransportChannel channel = mock(TransportChannel.class);
+                    doAnswer(invocation -> {
+                            try {
+                                // We expect to only send one response per request
+                                responseSetOnce.set((TransportResponse) invocation.getArguments()[0]);
+                                handleResponse(requestId, responseSetOnce.get());
+                                return null;
+                            } catch (Exception e) {
+                                throw new AssertionError(e);
+                            }
+                        }
+                    ).when(channel).sendResponse(any(TransportResponse.class));
+                    handlerRegistry.processMessageReceived(request, channel);
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                }
+            }
+        };
+
+        TestClusterNode(DiscoveryNode node, DeterministicTaskQueue deterministicTaskQueue) throws IOException {
```

Comment:
Setting up all the real services used for snapshotting with here with the exceptions:

* MockTransportService that just short-circuits the network.
* Mock ClusterStatePublisher that just short-circuits the network (I added a todo here, because I wasn't sure if we could maybe use the real thing here. It seemed very tricky to do so, but maybe it's not or worth the effort?)